website/addons/osfstorage/model.py

```diff
@@ -136,139 +102,131 @@ def create_waterbutler_log(self, auth, action, metadata):
+            Q('_id', 'eq', path) &
+            Q('node_settings', 'eq', node_settings)
+        )
 
     @classmethod
-    def find_by_path(cls, path, node_settings):
-        """Find a record by path and root settings record.
-
-        :param str path: Path to file or directory
-        :param node_settings: Root node settings record
-        """
-        try:
-            obj = cls.find_one(
-                Q('path', 'eq', path.rstrip('/')) &
-                Q('node_settings', 'eq', node_settings._id)
-            )
-            return obj
-        except modm_errors.NoResultsFound:
-            return None
+    def get_folder(cls, path, node_settings):
+        return cls.find_one(
+            Q('_id', 'eq', path) &
+            Q('kind', 'eq', 'folder') &
+            Q('node_settings', 'eq', node_settings)
+        )
 
     @classmethod
-    def get_or_create(cls, path, node_settings):
-        """Get or create a record by path and root settings record.
+    def get_file(cls, path, node_settings):
+        return cls.find_one(
+            Q('_id', 'eq', path) &
+            Q('kind', 'eq', 'file') &
+            Q('node_settings', 'eq', node_settings)
+        )
 
-        :param str path: Path to file or directory
-        :param node_settings: Root node settings record
-        :returns: Tuple of (record, created)
-        """
-        path = path.rstrip('/')
+    @property
+    @utils.must_be('folder')
+    def children(self):
+        return self.__class__.find(Q('parent', 'eq', self._id))
 
-        try:
-            obj = cls(path=path, node_settings=node_settings)
-            obj.save()
-        except KeyExistsException:
-            obj = cls.find_by_path(path, node_settings)
-            assert obj is not None
-            return obj, False
-
-        # Ensure all intermediate paths
-        if path:
-            parent_path, _ = os.path.split(path)
-            parent_class = cls.parent_class()
-            parent_obj, _ = parent_class.get_or_create(parent_path, node_settings)
-            parent_obj.append_child(obj)
-        else:
-            node_settings.file_tree = obj
-            node_settings.save()
-
-        return obj, True
+    @property
+    def is_folder(self):
+        return self.kind == 'folder'
 
-    def get_download_count(self, version=None):
-        """Return download count or `None` if this is not a file object (e.g. a
-        folder).
-        """
-        return None
+    @property
+    def is_file(self):
+        return self.kind == 'file'
 
-    def __repr__(self):
-        return '<{}(path={!r}, node_settings={!r})>'.format(
-            self.__class__.__name__,
-            self.path,
-            self.node_settings._id,
-        )
+    @property
+    def path(self):
+        return '/{}{}'.format(self._id, '/' if self.is_folder else '')
 
+    @property
+    def node(self):
+        return self.node_settings.owner
 
-class OsfStorageFileTree(BaseFileObject):
+    def materialized_path(self):
```

Comment:
Given that this may be expensive, would it make sense to make this a `cached_property`? You could use `werkzeug.utils.cached_property`? 

We may have to rethink the caching when we allow for file moving, but I think it's fine to do in the meantime.